lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,399 @@
+package networking
+
+import (
+	"context"
+	"encoding/json"
+	"log/slog"
+	"net"
+	"os"
+	"os/exec"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/sshutils/x11"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/uds"
+)
+
+// RequestBufferSize is the maximum amount of data we're comfortable writing at
+// once through a default unix datagram socket. Should fit comfortably in both
+// linux and darwin with default settings.
+const RequestBufferSize = 1024
+
+// Process represents an instance of a networking process.
+type Process struct {
+	// cmd is the running process command.
+	cmd *exec.Cmd
+	// conn is the socket used to request a dialer or listener in the process.
+	conn *net.UnixConn
+	// done signals when the process completes.
+	done chan struct{}
+	// killed is set to true when the process was killed forcibly.
+	killed atomic.Bool
+}
+
+// Request is a networking request.
+type Request struct {
+	// Operation is a networking operation.
+	Operation Operation
+	// Network is a network type.
+	Network string
+	// Address is a network address.
+	Address string
+	// X11Request contains additional info needed for x11 networking requests.
+	X11Request X11Request
+}
+
+// Operation is a networking operation.
+type Operation string
+
+const (
+	// NetworkingOperationDial is used to connect to a network address.
+	NetworkingOperationDial Operation = "dial"
+	// NetworkingOperationListen is used to create a local network listener.
+	NetworkingOperationListen Operation = "listen"
+	// NetworkingOperationListenAgent is used to create a local ssh-agent listener.
+	NetworkingOperationListenAgent Operation = "listen-agent"
+	// NetworkingOperationListenX11 is used to create a local x11 listener.
+	NetworkingOperationListenX11 Operation = "listen-x11"
+)
+
+// X11Config contains information used by the child process to set up X11 forwarding.
+type X11Request struct {
+	x11.ForwardRequestPayload
+	// DisplayOffset is the first display that we should try to get a unix socket for.
+	DisplayOffset int
+	// MaxDisplay is the last display that we should try to get a unix socket for, if all
+	// displays before it are taken.
+	MaxDisplay int
+}
+
+// NewProcess starts a new networking process with the given command, which should
+// be pre-configured from a ssh server context with Teleport reexec settings.
+func NewProcess(ctx context.Context, cmd *exec.Cmd) (*Process, error) {
+	// Create the socket to communicate over.
+	remoteConn, localConn, err := uds.NewSocketpair(uds.SocketTypeDatagram)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer remoteConn.Close()
+	remoteFD, err := remoteConn.File()
+	if err != nil {
+		localConn.Close()
+		return nil, trace.Wrap(err)
+	}
+	defer remoteFD.Close()
+	cmd.ExtraFiles = append(cmd.ExtraFiles, remoteFD)
+
+	// Propagate stderr from the spawned Teleport process to log any errors.
```

Comment:
If the child has stderr from the parent to log errors, why do we need the parent to listen on the control connection for errors?

Reply:
Logging in the child process just wouldn't working, I tried linking stdout in addition to stderr but no logs showed up. I also tested the logs in the exec process never show up. I couldn't figure out the root issue in a reasonable amount of time so I made this change and moved on.